docs/src/contributing/pgrx-internal.md

```diff
@@ -1 +1,112 @@
+- [correctness checks via the test suite](#test-suite)
+
+Though given that one component includes a pile of macros, and another component involves
+a rather sophisticated runtime element itself, the distinction often blurs.
+
+# runtime libraries
+
+## pgrx
+
+The main library that offers safe abstractions that wrap pgrx internals.
+
+### About macros and pgrx modules
+
+Currently, PGRX (the collection of libraries and tools) only assumes that this crate is used as
+the primary interface. Most pgrx code is generated via macros, which themselves may get used in
+odd situations. It is thus somewhat important to qualify most code that is introduced via macros
+by using `::pgrx` along with some module, so that the actual top-level crate named pgrx is used.
+
+Thus almost every library crate is named `pgrx-{module}` and also introduced as `pgrx::{module}`.
+
+### pgrx::prelude
+
+Previously, the recommended way to use pgrx was to glob-import everything. However, this has
+a problem in that pgrx has over time evolved to have use-cases which do not involve exposing
+all the `unsafe fn` of pgrx and its extensional innards to the "end-user".
+In other words, PL/Rust was born. Thus the prelude was introduced to limit our concerns.
+
+The prelude's contents should thus focus mostly on safe code and common types.
+
+### Designing new abstractions
+
+New abstractions in pgrx should avoid assuming that they are the only interface to Postgres, as
+pgrx-based extensions are often used in combination with C-based extensions, or are backed by
+considerable custom FFI written by a programmer.
+
+## pgrx-pg-sys
+
+This crate comes with an extensive `build.rs` to generate bindings for every Postgres version.
+
+## pgrx-macros
+
+...whether this is a runtime library or a build tool is a philosophical argument.
+
+# build tools
+
+## pgrx-sql-entity-graph
+
+This crate may seem... somewhat distressing. It is a rather lot of code, some of it without
+obvious purpose. One may develop fantasies about deleting it wholesale.  Unfortunately, it serves
+an irreplaceable functionality in pgrx, even if its role could be reduced.
+
+A primary task for pgrx, accomplished via cargo-pgrx and support libraries, is to generate SQL.
+This allows programmers to avoid separately maintaining the SQL that installs the extension when
+`CREATE EXTENSION` runs. This is not merely a convenience: misdeclaring extensions, functions,
+types, and operators can in some cases cause illogical behavior from Postgres. The query executor
+may assume something based on the SQL definitions, causing it to interact with extensions in ways
+that return invalid query results or, depending on the nature of the extension, corrupt a table's
+active rows or index. While Postgres employs various techniques to detect such incorrect behaviors
+and will abort transactions and roll back if it does, and this may not be "undefined behavior" in
+Rust terms, it is at least as equally undesirable!
```

Comment:
Find a way to work in the main citation for this, which is https://www.postgresql.org/docs/current/xoper-optimization.html